File: app/src/main/java/app/fynnjason/fivesingletondemo/java/DoubleCheckSingleton.java
package app.fynnjason.fivesingletondemo.java;

/**
 * Created by FynnJason on 2017/12/25.
 * 双重检测法，解决了懒加载和同步锁的缺点
 */

public class DoubleCheckSingleton {
    //第一步：私有构造
    private DoubleCheckSingleton() {

    }
    //第二步：创建静态对象
    private static volatile DoubleCheckSingleton INSTANCE;

    //第三步：对外公开获取方法
    public static DoubleCheckSingleton getInstance() {
        if (INSTANCE == null) {
            synchronized (DoubleCheckSingleton.class) {
                if (INSTANCE == null) {
                    INSTANCE = new DoubleCheckSingleton();
                }
            }
        }
        return INSTANCE;
    }
}

File: app/src/main/java/app/fynnjason/fivesingletondemo/kotlin/DoubleCheckSingleton.kt
package app.fynnjason.fivesingletondemo.kotlin

/**
 * Created by FynnJason on 2017/12/25.
 * 双重检测法，解决了懒加载和同步锁的缺点
 */
class DoubleCheckSingleton private constructor() {
    companion object {
        //Kotlin原生写法
        val INSTANCE_1 by lazy(mode = LazyThreadSafetyMode.SYNCHRONIZED) {
            DoubleCheckSingleton()
        }

        //翻译JAVA代码写法
        @Volatile
        private var INSTANCE_2: DoubleCheckSingleton? = null

        fun getInstance(): DoubleCheckSingleton {
            if (INSTANCE_2 == null) {
                synchronized(DoubleCheckSingleton::class) {
                    if (INSTANCE_2 == null) {
                        INSTANCE_2 = DoubleCheckSingleton()
                    }
                }
            }
            return INSTANCE_2!!
        }
    }
}